safehttp/form.go

```diff
@@ -0,0 +1,234 @@
+// limitations under the License.
+
+package safehttp
+
+import (
+	"errors"
+	"mime/multipart"
+	"strconv"
+)
+
+// Form contains parsed data either from URL field's query parameters or from
+// form parameters, part of the body of POST, PATCH or PUT requests that are not
+// multipart requests. The fields are only available after parsing the form,
+// through getter functions that specify the type.
+type Form struct {
+	values map[string][]string
+	err    error
+}
+
+// MultipartForm extends the Form structure to define a POST, PATCH or PUT
+// request that is multipart. Its fields are only available after parsing the
+// form, through getter functions that specify the type.
+type MultipartForm struct {
+	*Form
+	file map[string][]*multipart.FileHeader
+}
+
+// Int TODO
+func (f *Form) Int(paramName string, defaultValue int) int {
+	if f.err != nil {
+		return defaultValue
+	}
+	vals, ok := f.values[paramName]
+	if !ok {
+		f.err = errors.New("no value found for key " + paramName)
+		return defaultValue
+	}
+	paramVal, err := strconv.Atoi(vals[0])
+	if err != nil {
+		f.err = err
+		return defaultValue
+	}
+	return paramVal
+}
+
+// Uint TODO
+func (f *Form) Uint(paramName string, defaultValue uint64) uint64 {
+	if f.err != nil {
+		return defaultValue
+	}
+	vals, ok := f.values[paramName]
+	if !ok {
+		f.err = errors.New("no value found for key " + paramName)
+		return defaultValue
+	}
+	paramVal, err := strconv.ParseUint(vals[0], 10, 0)
+	if err != nil {
+		f.err = err
+		return defaultValue
+	}
+	return paramVal
+}
+func (f *Form) String(paramName string, defaultValue string) string {
+	if f.err != nil {
+		return defaultValue
+	}
+	vals, ok := f.values[paramName]
+	if !ok {
+		f.err = errors.New("no value found for key " + paramName)
+		return defaultValue
+	}
+	return vals[0]
+}
+
+// Float64 TODO
+func (f *Form) Float64(paramName string, defaultValue float64) float64 {
+	if f.err != nil {
+		return defaultValue
+	}
+	vals, ok := f.values[paramName]
+	if !ok {
+		f.err = errors.New("no value found for key " + paramName)
+		return defaultValue
+	}
+	paramVal, err := strconv.ParseFloat(vals[0], 64)
+	if err != nil {
+		f.err = err
+		return defaultValue
+	}
+	return paramVal
+}
+
+// Bool TODO
+func (f *Form) Bool(paramName string, defaultValue bool) bool {
+	if f.err != nil {
+		return defaultValue
+	}
+	vals, ok := f.values[paramName]
+	if !ok {
+		f.err = errors.New("no value found for key " + paramName)
+		return defaultValue
+	}
+	if vals[0] != "true" {
+		if vals[0] != "false" {
+			f.err = errors.New("values of form parameter " + paramName + " not a boolean")
+		}
+		return false
+	}
+	return true
+}
+
+// Slice TODO
+func (f *Form) Slice(slice interface{}, paramName string) {
+	mapVals, ok := f.values[paramName]
+	switch values := slice.(type) {
+	case *[]string:
+		if f.err != nil {
+			*values = nil
+			return
+		}
+		if !ok {
+			f.err = errors.New("no value found for key " + paramName)
+			*values = nil
+		}
```

Comment:
instead of having this check in every branch of the switch, can we have it before the switch?

Reply:
Implemented a solution in [d511d47](url)